clang/lib/CodeGen/CGDebugInfo.cpp

```diff
@@ -399,6 +399,18 @@ llvm::DIFile *CGDebugInfo::getOrCreateFile(SourceLocation Loc) {
 
   // Cache the results.
   auto It = DIFileCache.find(FileName.data());
+  if (SM.getFileEntryForID(SM.getMainFileID()) &&
```

Comment:
Can you please add comments to explain what is happening here? It also looks like there is code duplication from lines 405-411 and 424-430 and 567-571. Can this be refactored?